<a name=top></a><!doctype html><html lang=en-us><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><title>KJY</title><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css><link rel=stylesheet href=/css/style.css><link rel=stylesheet href=/css/fonts.css><link href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/github.min.css rel=stylesheet><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/highlight.min.js></script>
<script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/r.min.js></script>
<script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/bash.min.js></script>
<script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/python.min.js></script>
<script src=https://cdn.jsdelivr.net/npm/vega@5.17.0></script>
<script src=https://cdn.jsdelivr.net/npm/vega-lite@4.17.0></script>
<script src=https://cdn.jsdelivr.net/npm/vega-embed@6.12.2></script>
<script>hljs.initHighlightingOnLoad()</script></head><body><div class=wrapper><header class=header><nav class=nav><a href=/ class=nav-logo><img src=/media/%3cnil%3e width height alt></a><ul class=nav-links><li><a href=/cn/about/>关于</a></li><li><a href=/cn/posts/>日志</a></li><li><a href=/>English</a></li></ul></nav></header><main class=content role=main><div style=text-align:center><h1></h1><p></p><hr></div><span class=article-toolbar><a href=https://github.com/kongjianyang/kongjianyang.github.io/edit/master/content/cn/posts/2022-03-25-MACS2.md style=font-size:24px;color:#000 target=_blank><i class="fa fa-edit" aria-hidden=true title=编辑本页></i></a></span><div class="body-text list-text"><p>参考资料：</p><p><a href=http://www.360doc.com/content/18/0205/00/19913717_727772514.shtml target=_blank rel="noreferrer noopener">这可能是最棒的MACS2使用说明 (360doc.com)</a></p><p><a href=https://blog.csdn.net/weixin_43569478/article/details/108079812 target=_blank rel="noreferrer noopener">使用MACS2进行差异peak分析_生信修炼手册的博客-CSDN博客</a></p><p>MACS2作为使用最广泛的peak calling软件，在v2版本中添加了差异peak分析的功能，所有的子命令功能描述如下</p><p><img src="https://imgconvert.csdnimg.cn/aHR0cHM6Ly9tbWJpei5xcGljLmNuL21tYml6X3BuZy9LQXRSUEJCeWVHNVI5MGNRaWJxS2FibjVKM3ZwSDUyTWZnZUpkRTVwSGQza1BmdVNONEozMTVERVRKclpteFJxVGV0UXNQd3ZCVGdpY24zeFU3Tzdtd2hnLzY0MA?x-oss-process=image/format,png" alt></p><p>通过<code>bdgdiff</code>子命令来进行差异peak分析， 该命令不需要基于已有的peak calling结果，只需要输入每个样本对应的bedGraph格式的文件。需要注意的是，该命令只针对两个样本间的差异peak进行设计，适用于没有生物学重复的情况。</p><p>对于使用macs2来进行差异peak的完整流程，官方给出了详细的说明文档，链接如下</p><blockquote><p>https://github.com/taoliu/MACS/wiki/Call-differential-binding-events</p></blockquote><p>可以分为以下3步</p><h4 id=1-预测插入片段长度>1. 预测插入片段长度<a href=#1-预测插入片段长度 class=header-anchor arialabel=Anchor> #</a></h4><p>通过<code>predictd</code>子命令可以预测样本的fragment size，命令如下</p><pre tabindex=0><code>macs2 predictd -i input.bam
</code></pre><pre tabindex=0><code>#!/bin/bash

module load bioinfo
module load biocontainers/default
module load macs2/2.2.7.1-py39

for i in `ls [A-H][1-9][A-H][1-9][A-H][1-9].bam`;
do
msg=&#34;macs2 predictd -i $i&#34;;
echo $msg;
macs2 predictd -i $i;
</code></pre><h4 id=2-peak--calling>2. peak calling<a href=#2-peak--calling class=header-anchor arialabel=Anchor> #</a></h4><p>在peak calling时，需要添加<code>-B</code>参数，这样才可以输出样本对应的bedgraph文件，同时需要保证peak calling时采用一致的<code>--extsize</code>的值，就是第一步预测出来的数值，取多个样本的均值即可。官方也给出了推荐值，对于大多数的转录因子chip_seq数据，推荐值为200， 对于大部分组蛋白修饰的chip_seq数据，推荐值为147，命令如下</p><pre tabindex=0><code># condition1macs2 callpeak -B -t cond1_ChIP.bam -c cond1_Control.bam -n cond1 --nomodel --extsize 120# condition2macs2 callpeak -B -t cond1_ChIP.bam -c cond1_Control.bam -n cond1 --nomodel --extsize 120 --broad-cutoff 0.1
</code></pre><ul><li><code>-n/--name</code>表示实验的名字</li><li><code>-B/--bdg</code>: 以bedGraph格式存放<em>fragment pileup</em>, <em>control lambda</em>, <em>-log10pvalue</em> 和<em>log10qvale</em>.</li><li><code>-q</code>： q值(最小的FDR)的阈值，默认0.05。可以根据结果进行修正。q值是p值经Benjamini–Hochberg–Yekutieli修正后的值。</li><li><code>--broad--cutoff</code>: 用于过滤 <code>broad</code>得到的peak，默认是q值，如果设置 <code>-p</code>就用p值。</li><li><code>--nomodel</code>: 这个参数说明不需要MACS去构建模型，也就是说下面的参数除了 <code>--shift, --extsize</code>外都会被无视。</li><li><code>--extsize</code>： MACS使用这个参数将read以5&rsquo;-> 3&rsquo;衍生至等长片段。比如说你知道你的转录因子的结合区域是200bp，那么参数就是 <code>--extsize 200</code>。当且仅当 <code>--nomodel</code>和 <code>--fix-bimodal</code>设置使用。</li><li><code>-g</code>表示可比对的基因组大小。比如说人类是2.7e9，也就是2.7G，拟南芥根据NCBI显示是119,667,750，也就是1.2e8. 这个参数设置的原因，根据我读发表MACS的那篇文献推测，是因为MACS需要进行全基因组扫描构建双峰模型估计shift size。这个只是我猜测而已，反正根据自己物种来就对了。</li><li>-f/&ndash;format FORMAT<code>用来声明输入的文件格式，目前MACS能够识别的格式有 'ELAND', 'BED', 'ELANDMULTI', 'ELANDEXPORT', 'ELANDMULTIPET' (双端测序), 'SAM', 'BAM', 'BOWTIE', 'BAMPE', 'BEDPE'. 除'BAMPE', 'BEDPE'需要特别声明外，其他格式都可以用 </code>AUTO</li></ul><p>在运行这一步的时候，会输出每个样本过滤之后的reads数目，示意如下</p><pre tabindex=0><code># tags after filtering in treatment: 19291269# tags after filtering in control: 12914669
</code></pre><p>这个数值在差异分析中会用到，所以要记录下来。</p><h4 id=3-差异peak分析>3. 差异peak分析<a href=#3-差异peak分析 class=header-anchor arialabel=Anchor> #</a></h4><p>命令如下</p><pre tabindex=0><code>macs2 bdgdiff --t1 cond1_treat_pileup.bdg --c1 cond1_control_lambda.bdg --t2 cond2_treat_pileup.bdg \--c2 cond2_control_lambda.bdg --d1 12914669 --d2 14444786 -g 60 -l 120 --o-prefix diff_c1_vs_c2
</code></pre><p>其中<code>-d1</code>和<code>-d2</code>的值就是第二步运行时输出的reads数目，<code>-o</code>参数指定输出文件的前缀。运行成功后，会产生3个文件</p><ol><li><p>diff_c1_vs_c2_c3.0_cond1.bed</p></li><li><p>diff_c1_vs_c2_c3.0_cond2.bed</p></li><li><p>diff_c1_vs_c2_c3.0_common.bed</p></li></ol><p>其中, con1.bed保存了在condition1中上调的peak, con2.bed保存了在condition2中上调的peak, common.bed文件中保存的是没有达到阈值的，非显著差异peak。</p><p>上述3个文件格式是完全相同的，最后一列的内容为log10 likehood ratio值，用来衡量两个条件之间的差异，默认阈值为3，大于阈值的peak为组间差异显著的peak, 这个阈值可以通过<code>-c</code>参数进行调整。</p><h4 id=一些解释>一些解释<a href=#一些解释 class=header-anchor arialabel=Anchor> #</a></h4><p>This redundancy is consistently applied for both the ChIP and input samples.</p><blockquote><p><strong>Why worry about duplicates?</strong> Reads with the same start position are considered duplicates. These duplicates can arise from experimental artifacts, but can also contribute to genuine ChIP-signal.</p><ul><li><strong>The bad kind of duplicates:</strong> If initial starting material is low this can lead to overamplification of this material before sequencing. Any biases in PCR will compound this problem and can lead to artificially enriched regions. Also blacklisted (repeat) regions with ultra high signal will also be high in duplicates. Masking these regions prior to analysis can help remove this problem.</li><li><strong>The good kind of duplicates:</strong> Duplicates will also exist within highly efficient (or even inefficient ChIP) when deeply sequenced. Removal of these duplicates can lead to a saturation and so underestimation of the ChIP signal.</li></ul><p><strong>The take-home:</strong> Consider your enrichment efficiency and sequencing depth. But, because we cannot distinguish between the good and the bad, best practice is to remove duplicates prior to peak calling. Retain duplicates for differential binding analysis. Also, if you are expecting binding in repetitive regions keep duplicates and multiple mappers.</p></blockquote><p style=color:#777>最后一次修改于 0001-01-01</p></div><a href=#top><i class="fa fa-chevron-up" style=font-size:30px;color:#000></i></a></main><footer class=footer><script src=https://utteranc.es/client.js repo=kongjianyang/kongjianyang.github.io issue-term=pathname theme=github-light crossorigin=anonymous async></script><script type=text/javascript src=/js/math-code.js></script>
<script async src="//mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script type=text/javascript src=/js/center-img.js></script><ul class=footer-links><li><a href=/cn/posts/index.xml type=application/rss+xml title="RSS feed">订阅</a></li><li><a href=http://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh target=_blank>版权
<i class="fa fa-cc" aria-hidden=true title="Attribution-NonCommercial-ShareAlike 4.0 International"></i></a></li></ul><div class=copyright-text>©
KJY
2018-2021</div></footer>